src/flatten.jl

```diff
@@ -0,0 +1,101 @@
+
+isassociative(x::NamedTuple) = true
+isassociative(x::AbstractDict) = true
+isassociative(x::Iterators.Pairs) = true
+isassociative(x::Vector{<:Pair}) = true
+isassociative(x) = false
+
+# NOTE: NamedTuples only support symbol names, so we use a simple :_ delimiter
+function flatten(x::NamedTuple; delim=DEFAULT_FLATTEN_DELIM)::NamedTuple
+    kwargs = map(flatten(pairs(x))) do (k, v)
+        _k = isa(k, Tuple) ? join(k, delim) : k
+        return Symbol(_k) => v
+    end
+    return (; kwargs...)
+end
+
+# Since the key type could be changed as part of flattening we can only ensure the that
+# result is a `Dict` (vs the same input dict).
+function flatten(x::AbstractDict; delim=nothing)
+    return Dict(flatten(collect(x); delim=delim)...)
+end
+
+function flatten(x::Union{Vector{<:Pair}, Iterators.Pairs}; delim=nothing)
+    result = Channel{Pair}() do chnl
```

Comment:
Sure. It's a similar use case to `walkdir` in base julia or `s3_list_objects` in AWSS3.jl, where we want to recursively build up an iterator. `Channel`s are nice for this because we can describe a function that takes a channel and puts arbitrary items into it, without needing to know the size ahead of time.

https://github.com/JuliaLang/julia/blob/df7334a25e38b42e9fb2808d4eaac2ccb9d2d03a/base/file.jl#L898
https://github.com/JuliaCloud/AWSS3.jl/blob/54cddc966996bb2d888b23f234c05c48d74e006d/src/AWSS3.jl#L408

In this particular case, we're:

1. Constructing a Channel{Pair} because our input to the function is a vector or iterator of pairs.
2. In our channel function we're iterating through each pair and either:
  - Recursively calling `flatten` if the value of a pair is itself associative (as describe in the methods above), and then putting those child pairs into our channel
  - Directly inserting the pair into our channel
  
  Does that make sense?

Reply:
ah right, yeah that does make sense now thanks. This bit was the key thing for my understanding

>we can describe a function puts arbitrary items into [a channel], without needing to know the size ahead of time.